pkg/operator/controllers/machine/machine.go

```diff
@@ -37,9 +37,15 @@ func (r *Reconciler) machineValid(ctx context.Context, machine *machinev1beta1.M
 		return []error{fmt.Errorf("machine %s: provider spec missing", machine.Name)}
 	}
 
-	machineProviderSpec, err := utilmachine.UnmarshalAzureProviderSpec(machine.Name, utilmachine.Machine, machine.Spec.ProviderSpec.Value.Raw)
+	o, _, err := scheme.Codecs.UniversalDeserializer().Decode(machine.Spec.ProviderSpec.Value.Raw, nil, nil)
 	if err != nil {
-		errs = append(errs, err)
+		return []error{err}
+	}
+	machineProviderSpec, ok := o.(*machinev1beta1.AzureMachineProviderSpec)
+	if !ok {
+		// This should never happen: codecs uses scheme that has only one registered type
+		// and if something is wrong with the provider spec - decoding should fail
+		return []error{fmt.Errorf("machine %s: failed to read provider spec: %T", machine.Name, o)}
```

Comment:
The thing is - we need to support both types: old one (we run the operator in the older clusters) and the new one. That's why we introduced `utilmachine.UnmarshalAzureProviderSpec`.

Note: it is very likely that there is no better solution than having `utilmachine.UnmarshalAzureProviderSpec`, but I would like you to explore the options.

I was hoping that we could somehow solve it on the scheme level instead of having `utilmachine.UnmarshalAzureProviderSpec` which does relatively unreliable check of `strings.Contains(string(rawProviderSpec), azureProviderSpec)`. Maybe by registring `azureproviderconfig.openshift.io` to be unmarshalled into the new type? Or somehow else?

I'm wondering what other repos in https://github.com/openshift do to deal with this migration and backward compatibility. We can ask around: `#forum-cloud` on slack is probably a good place to start.

Reply:
do you know in what field is that `azureproviderspec` string supposed to show up ?

Reply:
@m1kola I've updated a scheme. Could you take a look at it, please? Also, I am wondering if we need some tests to be added.

Reply:
@yehornaumenko yes, I think we need to update few tests to make sure that we unmarshall both old and new type correctly. I think we need to update tests for the following code:

* `pkg/cluster/fixmcsuserdata.go`
* `pkg/operator/controllers/machine/machine.go`
* `pkg/util/clusterdata/worker_profiles_task.go`
* `pkg/util/subnet/cluster_subnets.go`
* Something else?